File: 081825/longestsubstring.py
def length_of_longest_substring(s: str) -> int:
    chars = set()  # to track unique characters in current window
    left = 0
    max_len = 0

    for right, c in enumerate(s):
        # if duplicate, shrink window from the left
        while c in chars:
            chars.remove(s[left])
            left += 1
        
        chars.add(c)
        max_len = max(max_len, right - left + 1)

    return max_len

# --- Example usage ---
s = "abcabcbb"
print("Longest substring length:", length_of_longest_substring(s))  # Output: 3

File: August/18 08 25/longestsubstring.cpp
#include <iostream>
#include <string>
#include <unordered_set>
using namespace std;

int lengthOfLongestSubstring(string s) {
    unordered_set<char> chars; // to track unique characters in current window
    int left = 0, maxLen = 0;

    for (int right = 0; right < s.length(); right++) {
        // if duplicate, shrink window from left
        while (chars.count(s[right])) {
            chars.erase(s[left]);
            left++;
        }

        chars.insert(s[right]);
        maxLen = max(maxLen, right - left + 1);
    }

    return maxLen;
}

// --- Example usage ---
int main() {
    string s = "abcabcbb";
    cout << "Longest substring length: " << lengthOfLongestSubstring(s) << endl;
    return 0;
}